### PR TITLE
exception handling for promise

### PR DESCRIPTION
**Pull request by sonnyp**
related conversation https://www.reddit.com/r/javascript/comments/4eqhic/make_a_javascript_function_support_both_promise/d22slpz

### CODE REVIEW

**Comment by tunnckoCore** [comment]
Looks okey to me, but I don't think you should check err for null and undefined.

**Comment by sonnyp** [comment]
@tunnckoCore what would you check for?

**Comment by tunnckoCore** [comment]
Generally we should not check for nothing - just `if (err) reject(err)`. But i'm using `is-typeof-error`, because throwing different than error object is bad practice.

**Comment by sonnyp** [comment]
so you mean just checking if the value is truthy

Yeah, wasn't sure about that.

but if you call `cb(foo)` and foo is for some reason zero or en empty string, I'd still expect the promise to reject with that value

**Comment by tunnckoCore** [comment]
It's just not our job. And that's totally wrong and we should not care about and should not support such things. Community know, community follow conventions and best practices. It's not normal user to expect resolved promise when pass first argument to callback, it's not natural and it's anti-node - we should not support such things. Such user should go back and learn how node (and javascript in general) works and why it works in exactly this way.

**Comment by sonnyp** [comment]
>  It's not normal user to expect resolved promise when pass first argument to callback

I don't think you understood me, if the callback is called with a first argument, the promise is rejected (except if that argument is null or undefined)

**Comment by tunnckoCore** [comment]
I think I perfectly understood you first time. What you expect when you pass `0` or such values as first argument? Resolved promise with `0` as result? That's totally wrong. And that's what I'm talking about, the all world is driven by conventions and know what the falsey value is and how to use callbacks.

**Comment by sonnyp** [comment]
> What you expect when you pass 0 or such values as first argument? Resolved promise with 0 as result?

No, if you pass 0 as first value the promise reject with that value.

**Comment by tunnckoCore** [comment]
Exactly. What's the problem? What wrong you see? In general case, yea, it's wrong but we should not care and all is conventions. That's why I use and more reasonable check will be `err instanceof Error`.

**Comment by benjamingr** [comment]
So, some thoughts:
- The proposal for Node didn't get shut down. It's still going.
- The particular dual syntax was "shut down" because of APIs with optional arguments like `crypto` and other similar APIs, where there is ambiguity - as well as functions with non-conforming callback signatures like `fs.exists`.
- promise returning functions should _never_ throw, only reject. This is what promise users expect - though it becomes a non-issue when async/await lands.

**Comment by benjamingr** [comment]
Generally, I am in favor of rejecting on errors rather than throwing. A function should always or never behave asynchronously - throwing/returning values synchronously is a sham - this is also true with callbacks - see http://blog.izs.me/post/59142742143/designing-apis-for-asynchrony.

Note that this PR is not needed and can be safely archived - this is what the promise constructor does anyway... when you do `new P` and pass a function it converts exceptions to rejections, why? To guard against this very case.

``` js
new Promise(() => {
    throw new Error();
}).catch(e) {
    // the error ends up here, although it is synchronous.
});

```

File: index.js [hunk 1]
```diff
@@ -9,13 +9,18 @@
     } else {
       var P = Promise || global.Promise
```
Comment (benjamingr) [inline index.js]: doesn't this line break if `Promise` is not declared? (Also in master) Why would it be declared but falsey?

**Comment by sonnyp** [comment]
> Note that this PR is not needed and can be safely archived

Indeed, I modified the PR to only includes the tests and removed changes from `index.js`

**Comment by benjamingr** [comment]
LGTM

**Comment by sonnyp** [comment]
@benjamingr thanks

**Pull request merged**